### PR TITLE
Invalidate cache after write operation

### LINKED ISSUE #13186: Enable invalidating entries on client-side metadata cache when new files are generated
**Summary**
Motivated by https://github.com/Alluxio/alluxio/issues/13129 

When `alluxio.user.metadata.cache.enabled=true`, adding a new file to alluxio path `/foo` should invalidate corresponding metadata entries cached locally at `alluxio.client.file.MetadataCachingBaseFileSystem`.

**Urgency**
Medium

### PR DESCRIPTION
**Pull request by maobaolong**
Fix #13186

### CODE REVIEW

**Comment by maobaolong** [comment]
Hi @apc999 Please take a look at this PR, thanks.

**Review by apc999 — CHANGES_REQUESTED** [review]
@maobaolong thanks for the improvement. The code looks good to me. can you

1. update PR description, so we can see more details in git commit history
2. add unit test of invalidate?

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCache.java [hunk 1]
```diff
@@ -150,6 +150,15 @@ public void put(AlluxioURI dir, List<URIStatus> statuses) {
     return null;
   }
 
+  /**
+   * Invalidate the cache of path.
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCache.java:154]: nit:
```suggestion
   * Invalidates the cache of path.
```

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 2]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
...
+      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException {
+    super.createDirectory(path, options);
+    mMetadataCache.invalidate(path.getParent());
+    mMetadataCache.invalidate(path);
+  }
+
+  @Override
+  public FileOutStream createFile(AlluxioURI path)
+      throws IOException,
+      AlluxioException {
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:93]:
```suggestion
      throws IOException, AlluxioException {
```

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 3]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
...
+    FileOutStream outStream = super.createFile(path);
+    mMetadataCache.invalidate(path.getParent());
+    mMetadataCache.invalidate(path);
+    return outStream;
+  }
+
+  @Override
+  public FileOutStream createFile(AlluxioURI path, CreateFilePOptions options)
+      throws IOException,
+      AlluxioException {
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:103]: can you fix the format and following?

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 4]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
...
+    FileOutStream outStream = super.createFile(path, options);
+    mMetadataCache.invalidate(path.getParent());
+    mMetadataCache.invalidate(path);
+    return outStream;
+  }
+
+  @Override
+  public void delete(AlluxioURI path)
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:111]: override this method, which is the default impl of `delete`
```suggestion
  rename(AlluxioURI src, AlluxioURI dst, RenamePOptions options)
```

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 5]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
...
+      throws IOException,
+      AlluxioException {
+    super.delete(path);
+    mMetadataCache.invalidate(path.getParent());
+    mMetadataCache.invalidate(path);
+  }
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:117]: default impl of `delete(AlluxioURI)` is based on `delete(AlluxioURI, DeletePOptions)`. 
So you don't need both

```suggestion
  public void delete(AlluxioURI path)
      throws IOException,
      AlluxioException {
    super.delete(path);
    mMetadataCache.invalidate(path.getParent());
    mMetadataCache.invalidate(path);
  }
```

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 6]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
         TimeUnit.SECONDS, new SynchronousQueue<>());
   }
 
+  @Override
+  public void createDirectory(AlluxioURI path, CreateDirectoryPOptions options)
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:83]: add `public void createDirectory(AlluxioURI path)` too?

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 7]
```diff
@@ -73,12 +79,78 @@ public MetadataCachingBaseFileSystem(FileSystemContext context) {
...
+      throws FileAlreadyExistsException, InvalidPathException, IOException, AlluxioException {
+    super.createDirectory(path, options);
+    mMetadataCache.invalidate(path.getParent());
+    mMetadataCache.invalidate(path);
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:87]: shall we invalidate the cache first before modifying file system, to be conservative?
```suggestion
    mMetadataCache.invalidate(path.getParent());
    mMetadataCache.invalidate(path);
    super.createDirectory(path, options);

```

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java (hunk 6 above)
Comment (maobaolong) [reply core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:92]: Sure

**Comment by maobaolong** [comment]
@apc999 Thanks for the review and suggestion, addressed your comments except the adding ut request, will add it soon

File: tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java [hunk 8]
```diff
@@ -236,7 +236,7 @@ public void basicRenameTest7() throws Exception {
     // Due to Hadoop 1 support we stick with the deprecated version. If we drop support for it
     // FSDataOutputStream.hflush will be the new one.
     //#ifdef HADOOP1
-    o.sync();
+//    o.sync();
```
Comment (apc999) [inline tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java:239]: revert change to this file?

**Comment by apc999** [comment]
Can you revert the change in `tests/src/test/java/alluxio/client/hadoop/FileSystemRenameIntegrationTest.java`? Otherwise, LGTM

**Comment by maobaolong** [comment]
@apc999 Thanks for pointing out my bad, I revert the accidental changes, PTAL

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Pull request merged**